pkg/kv/kvserver/rangefeed/processor.go

```diff
@@ -177,10 +178,107 @@ func NewProcessor(cfg Config) *Processor {
 	}
 }
 
+// A CatchupIterator is an iterator for catchup-scans.
+type CatchupIterator interface {
+	storage.SimpleMVCCIterator
+
+	UnsafePrevValueForKey(storage.MVCCKey) ([]byte, error)
+}
+
 // IteratorConstructor is used to construct an iterator. It should be called
 // from underneath a stopper task to ensure that the engine has not been closed.
 type IteratorConstructor func() storage.SimpleMVCCIterator
 
+// CatchupIteratorConstructor is used to construct an iterator that
+// can be used for catchup-scans. It should be called from underneath
+// a stopper task to ensure that the engine has not been closed.
+type CatchupIteratorConstructor func() CatchupIterator
+
+type catchupIterator struct {
+	storage.SimpleMVCCIterator
+
+	// prevValIter is used to look up the previous values of keys
+	// when using the TBI optimization, since the previous value
+	// of the key is likely outside the time bounds of the TBI.
+	//
+	// TODO(ssd): It should be possible to use the underlying
+	// non-timebound iterator in IncrementalIterator for this.
+	prevValIter storage.SimpleMVCCIterator
+	close       func()
+}
+
+// NewCatchupIterator returns a CatchupIterator for the given Reader.
+// If useTBI is true, a time-bound iterator will be used, configured
+// with a start time taken from the RangeFeedRequest.
+func NewCatchupIterator(
+	reader storage.Reader, args *roachpb.RangeFeedRequest, useTBI bool, closer func(),
+) CatchupIterator {
+	ret := catchupIterator{
+		close: closer,
+	}
+	if useTBI {
+		ret.SimpleMVCCIterator = storage.NewMVCCIncrementalIterator(reader, storage.MVCCIncrementalIterOptions{
+			EnableTimeBoundIteratorOptimization: true,
+			EndKey:                              args.Span.EndKey,
+			// StartTime is exclusive
+			StartTime:    args.Timestamp.Prev(),
+			EndTime:      hlc.MaxTimestamp,
+			IntentPolicy: storage.MVCCIncrementalIterIntentPolicyEmit,
+			InlinePolicy: storage.MVCCIncrementalIterInlinePolicyEmit,
+		})
+		if args.WithDiff {
+			ret.prevValIter = reader.NewMVCCIterator(storage.MVCCKeyAndIntentsIterKind, storage.IterOptions{
```

Comment:
Because our `reader` here will be an Engine, the comments in the code lead me to believe that we might have a problem here with inconsistent iterators.  Need to look into this more.